asdf/block.py

```diff
@@ -78,8 +78,10 @@ def _add(self, block):
             msg = "Can not add second streaming block"
             raise ValueError(msg)
 
-        if block._data is not None:
-            self._data_to_block_mapping[id(block._data)] = block
+        if block._data is not None or key is not None:
+            if key is None:
+                key = id(block._data)
+            self._data_to_block_mapping[key] = block
```

Comment:
Does this imply that if created with a key, that the id cannot be used as a key for that block? May that cause problems with future code that presumes that id should always work?

Reply:
Yes that is correct. If a `key` is provided then `id(block._data)` will not be used to generate a `key` and an attempt to get a block using `id(block._data)` will fail (if it does not match the `key`).

Is there a current (or future) use that you think might conflict with this? I'm expecting that when we update ndarray to a converter that we will continue to use `id(block._data)` (equal to `id(base_array)`). The `key` is more useful for objects where the data might not have a consistent location in memory (such as zarr, which reloads the data on each read, creating a new id, and instead a key based on the chunk key is more consistent). Here's an example usage in the `asdf-zarr` prototype: https://github.com/braingram/asdf_zarr/blob/deferred_block/src/asdf_zarr/converter.py#L143